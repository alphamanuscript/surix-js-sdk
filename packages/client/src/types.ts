import { WrappedEntity, WrappedFile } from '@surix/data-helpers';

export interface Project {
  entities: ProjectEntities;
  files: ProjectFiles;
  tags: ProjectTags;
}

export interface ProjectEntities {
  get(id: string): Promise<WrappedEntity>;
  query(query?: Query): Promise<WrappedEntity[]>;
  create(entity: Entity): Promise<WrappedEntity>;
  put(entity: Entity): Promise<WrappedEntity>;
}

export interface ProjectFiles {
  get(id: string): Promise<WrappedFile>;
  list(): Promise<WrappedFile[]>;
}

export interface ProjectTags {
  list(): Promise<TagList>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];

export enum FieldType {
  TEXT = 'text',
  BOOLEAN = 'boolean',
  NUMBER = 'number',
  DATETIME = 'datetime',
  PHONE = 'phone',
  EMAIL = 'email',
  FILE = 'file',
  OBJECT = 'object',
  LIST = 'list'
}

export interface FieldPair {
  type: FieldType;
  value: RawEntityData | string | RawEntityData[] | number;
}

export interface RawEntityField extends FieldPair {
}

export interface RawEntityData {
  [fieldName: string]: RawEntityField;
}

/**
 * this interface describe the shape of the entity
 * as used within the system and the API
 */
export interface RawEntity {
  data: RawEntityData;
  tags?: string[];
}

export interface DataField {
  [fieldName: string]: string | number | object | boolean;
}
export interface Entity {
  data: DataField;
  tags?: string[];
  _id?: string;
}