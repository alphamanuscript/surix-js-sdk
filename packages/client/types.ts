import { EntityWrapper } from '@surix/data-helpers';

export interface Project {
  db: ProjectDatabase;
  files: ProjectFiles;
}

export interface ProjectDatabase {
  get(id: string): Promise<EntityWrapper>;
  query(query: Query): Promise<EntityWrapper[]>;
  getTags(): Promise<TagList[]>;
}

export interface ProjectFiles {
  get(id: string): Promise<ApiFile>;
  list(): Promise<ApiFile[]>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface ApiFile {
  name: string;
  status: 'failed' | 'ready' | 'pending';
  public: boolean;
  downloadUrl: string;
  tags: TagList[];
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];