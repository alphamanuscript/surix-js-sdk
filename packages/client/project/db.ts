import { ApiEntity, EntityWrapper } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectDatabase, Query, TagList } from '../types';

export function getProjectDb (projectId: string, apiClient: AxiosInstance): ProjectDatabase {
  return {
    async get (entityId: string): Promise<EntityWrapper> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return new EntityWrapper(res.data);
    },
    async query(query: Query): Promise<EntityWrapper[]> {
      const res = await apiClient.post<ApiEntity[]>(`/projects/${projectId}/entities/query`, query);
      const entities = res.data;
      return entities.map(entity => new EntityWrapper(entity));
    },
    async getTags (): Promise<TagList[]> {
      const res = await apiClient.get<TagList[]>(`/projects/${projectId}/tags`);
      return res.data;
    }
  };
}