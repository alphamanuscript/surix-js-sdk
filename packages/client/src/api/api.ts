import axios, { AxiosInstance } from 'axios';

export function getApiClient (baseUrl: string, apiKey: string): AxiosInstance {
  const axiosInstance = axios.create({
    baseURL: baseUrl
  })
  axiosInstance.defaults.headers['Authorization'] = `apikey ${apiKey}`
  return axiosInstance
}