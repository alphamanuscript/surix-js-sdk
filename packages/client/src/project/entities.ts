import { 
  ApiEntity,  
  expandEntity, 
  expandQuery, 
  wrapEntity, 
  wrapEntityArray, 
  WrappedEntity
} from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import {  DeletedEntities, Entity, ProjectEntities, Query } from '../types';

export function getProjectEntities (projectId: string, apiClient: AxiosInstance): ProjectEntities {
  return {
    async get(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return wrapEntity(res.data);
    },
    async query(query: Query = {}): Promise<WrappedEntity[]> {
      const expandedQuery = expandQuery(query);
      if (Object.keys(expandedQuery.query).length === 0) {
        delete expandedQuery.query;
      }
      const res = await apiClient.post<ApiEntity[]>(
        `/projects/${projectId}/entities/query`, expandedQuery);
      const entities = res.data;
      return wrapEntityArray(entities);
    },
    async update(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity };
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.patch<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async put(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity }; // Make a shallow copy
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.put<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async delete(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.delete(`/projects/${projectId}/entities/${entityId}`);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async deleteMany(entityIds: string[]): Promise<DeletedEntities> {
      const res = await apiClient.delete(`/projects/${projectId}/entities`, 
      {
        data: {
          entities: entityIds
        }
      });
      return res.data;
    },
    async create(entity: Entity): Promise<WrappedEntity> {
      const expandedEntity = expandEntity(entity);
      const res = await apiClient.post<ApiEntity>(
        `/projects/${projectId}/entities`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    }
  };
}