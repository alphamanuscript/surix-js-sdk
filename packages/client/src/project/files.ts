import {
  ApiFile,
  wrapFile,
  wrapFileArray,
  WrappedFile } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectFiles} from '../types';

export function getProjectFiles (projectId: string, apiClient: AxiosInstance): ProjectFiles {
  return {
    async get (fileId: string): Promise<WrappedFile> {
      const res = await apiClient.get<ApiFile>(`/projects/${projectId}/files/${fileId}`);
      return wrapFile(res.data);
    },
    async list (): Promise<WrappedFile[]> {
      const res = await apiClient.get<ApiFile[]>(`/projects/${projectId}/files`);
      return wrapFileArray(res.data);
    }
  };
}