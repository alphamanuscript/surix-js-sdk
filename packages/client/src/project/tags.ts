import { AxiosInstance } from 'axios';
import { ProjectTags, Tag, TagList  } from '../types';

export function getProjectTags (projectId: string, apiClient: AxiosInstance): ProjectTags {
  return {
    async list (): Promise<TagList> {
      const res = await apiClient.get<TagList>(`/projects/${projectId}/tags`);
      return res.data;
    }, 
    async rename (oldName: string, newName: string): Promise<Tag> {
      const res = await apiClient.put<Tag>(`/projects/${projectId}/tags/${oldName}`, {
        name: newName
      });
      return res.data;
    }
  };
}