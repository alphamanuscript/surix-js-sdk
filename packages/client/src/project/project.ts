import { AxiosInstance } from 'axios';
import { getProjectDb } from './db';
import { getProjectFiles } from './files';

import { Project } from '../types';

export function getProjectApi (projectId: string, apiClient: AxiosInstance): Project {
  return {
    db: getProjectDb(projectId, apiClient),
    files: getProjectFiles(projectId, apiClient)
  };
}