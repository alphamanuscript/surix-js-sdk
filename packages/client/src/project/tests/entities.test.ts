import * as dataHelpers from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import * as api from '../../api';
import { getProjectApi } from '../project';

import { Query } from '../../types';

describe('Project Entities', () => {
  const apiEntities = [
    {
      _id: 'e1',
      tags: ['t1'],
      createdBy: { _id: 'u1', type: 'user' },
      createdAt: '2017-02-22T09:19:33.885Z',
      updatedAt: '2018-10-22T09:19:33.885Z',
      data: {
        name: {
          type: 'text',
          value: 'Foo',
          name: 'name',
          label: 'Label'
        }
      }
    },
    {
      _id: 'e2',
      tags: ['t2'],
      createdBy: { _id: 'u2', type: 'user' },
      createdAt: '2017-02-23T09:19:33.885Z',
      updatedAt: '2018-10-23T09:19:33.885Z',
      data: {
        name: {
          type: 'text',
          value: 'Bar',
          name: 'name',
          label: 'Label'
        }
      }
    }
  ];
  describe('get', () => {
    let apiClient: AxiosInstance;
    async function callMockedGet (): Promise<dataHelpers.WrappedEntity> {
      apiClient = api.getApiClient('http://baseurl');
      jest.spyOn(apiClient, 'get').mockReturnValue(Promise.resolve({ data: apiEntities[0] }));
      jest.spyOn(dataHelpers, 'wrapEntity');
      const project = getProjectApi('project1', apiClient);
      const entity = await project.entities.get('e1');
      return entity;
    }
    it('should call the GET /project/:pid/entities/:id endpoint', async () => {
      await callMockedGet();
      expect(apiClient.get).toHaveBeenCalledWith(`/projects/project1/entities/e1`);
    });
    it('should return the entity transformed to WrappedEntity', async () => {
      const entity = await callMockedGet();
      expect(dataHelpers.wrapEntity).toHaveBeenCalledWith(apiEntities[0]);
      expect(entity).toMatchSnapshot();
    });
  });
  describe('query', () => {
    let apiClient: AxiosInstance;
    async function callMockedQuery (query?: Query): Promise<dataHelpers.WrappedEntity[]> {
      apiClient = api.getApiClient('http://baseurl');
      jest.spyOn(apiClient, 'post').mockReturnValue(Promise.resolve({ data: apiEntities }));
      jest.spyOn(dataHelpers, 'wrapEntityArray');
      const project = getProjectApi('project1', apiClient);
      const entities = await project.entities.query(query);
      return entities;
    }
    it('should call the GET /project/:pid/entities/query endpoint with query', async () => {
      const query = { limit: 10 };
      await callMockedQuery(query);
      expect(apiClient.post).toHaveBeenCalledWith(`/projects/project1/entities/query`, query);
    });
    it('should return the entities transformed to array of WrappedEntity', async () => {
      const query = { limit: 10 };
      const entities = await callMockedQuery(query);
      expect(dataHelpers.wrapEntityArray).toHaveBeenCalledWith(apiEntities);
      expect(entities).toMatchSnapshot();
    });
  
    describe('when query is not provided', () => {
      it('should call query endpoint with empty object', async () => {
        await callMockedQuery();
        expect(apiClient.post).toHaveBeenCalledWith(`/projects/project1/entities/query`, {});
      });
    });
  }); 
});