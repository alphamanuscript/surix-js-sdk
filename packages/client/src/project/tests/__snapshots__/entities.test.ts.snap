// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Entities get should return the entity transformed to WrappedEntity 1`] = `
t {
  "_entity": Object {
    "_id": "e1",
    "createdAt": "2017-02-22T09:19:33.885Z",
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "data": Object {
      "name": Object {
        "label": "Label",
        "name": "name",
        "type": "text",
        "value": "Foo",
      },
    },
    "tags": Array [
      "t1",
    ],
    "updatedAt": "2018-10-22T09:19:33.885Z",
  },
  "createdAt": 2017-02-22T09:19:33.885Z,
  "updatedAt": 2018-10-22T09:19:33.885Z,
}
`;

exports[`Project Entities query should return the entities transformed to array of WrappedEntity 1`] = `
Array [
  t {
    "_entity": Object {
      "_id": "e1",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "data": Object {
        "name": Object {
          "label": "Label",
          "name": "name",
          "type": "text",
          "value": "Foo",
        },
      },
      "tags": Array [
        "t1",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
    },
    "createdAt": 2017-02-22T09:19:33.885Z,
    "updatedAt": 2018-10-22T09:19:33.885Z,
  },
  t {
    "_entity": Object {
      "_id": "e2",
      "createdAt": "2017-02-23T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u2",
        "type": "user",
      },
      "data": Object {
        "name": Object {
          "label": "Label",
          "name": "name",
          "type": "text",
          "value": "Bar",
        },
      },
      "tags": Array [
        "t2",
      ],
      "updatedAt": "2018-10-23T09:19:33.885Z",
    },
    "createdAt": 2017-02-23T09:19:33.885Z,
    "updatedAt": 2018-10-23T09:19:33.885Z,
  },
]
`;
